api/authentication.py

```diff
@@ -1,33 +1,353 @@
+        if not self.from_cache and self.error in self._log_failure:
+            logger.error(
+                "accounts_introspection_failed",
+                extra={
+                    "error": self.error,
+                    "error_args": [shlex.quote(arg) for arg in self.error_args],
+                    "status_code": self.status_code,
+                    "data": self.data,
+                    "method": method,
+                    "path": path,
+                    "introspection_time_s": self.request_s,
+                },
+            )
+        code = self._exception_code[self.error]
+        if code == 401:
+            raise IntrospectAuthenticationFailed(self)
+        elif code == 503:
+            raise IntrospectUnavailable(self)
+
+    def as_cache_value(self) -> CachedFxaIntrospectResponse:
+        cached: CachedFxaIntrospectResponse = {"error": self.error}
+        if self.status_code:
+            cached["status_code"] = self.status_code
+        if self.data:
+            cached["data"] = self.data
+        if self.error_args:
+            cached["error_args"] = self.error_args
+        return cached
+
+    def save_to_cache(self, cache: BaseCache, token: str, timeout: int) -> None:
+        cache.set(get_cache_key(token), self.as_cache_value(), timeout)
+
+
+class IntrospectUnavailable(APIException):
+    status_code = 503
+    default_detail = "Introspection temporarily unavailable, try again later."
+    default_code = "introspection_service_unavailable"
+
+    def __init__(
+        self, introspection_error: IntrospectionError, *args: Any, **kwargs: Any
+    ) -> None:
+        self.introspection_error = introspection_error
+        super().__init__(*args, **kwargs)
+
+
+class IntrospectAuthenticationFailed(AuthenticationFailed):
+    def __init__(
+        self, introspection_error: IntrospectionError, *args: Any, **kwargs: Any
+    ) -> None:
+        self.introspection_error = introspection_error
+        super().__init__(*args, **kwargs)
+
+
+def load_introspection_result_from_cache(
+    cache: BaseCache, token: str
+) -> IntrospectionResponse | IntrospectionError | None:
+    cache_key = get_cache_key(token)
+    cached = cache.get(cache_key)
+    if cached is None or not isinstance(cached, dict):
+        return None
+    data_maybe = cached.get("data")
+    if error := cached.get("error"):
+        return IntrospectionError(
+            token,
+            error=error,
+            status_code=cached.get("status_code"),
+            data=data_maybe,
+            error_args=cached.get("error_args"),
+            from_cache=True,
+        )
+    response = IntrospectionResponse(token, data=data_maybe, from_cache=True)
+    if response.is_expired:
+        return IntrospectionError(
+            token,
+            "TokenExpired",
+            error_args=[str(response.time_to_expire)],
+            data=data_maybe,
+            from_cache=True,
+        )
```

Comment:
question (blocking just so I don't forget to check into this before we merge): In previous (2024?) code, I recall we tried to re-fetch a new un-expired token in the case where the current token is expired. Can we still do that? Or maybe we're already doing that in the Firefox code?

Reply:
Okay yeah that sounds good. Thanks for checking into that!